synapse/lib/json.py

```diff
@@ -0,0 +1,204 @@
+
+    Arguments:
+        s (str | bytes): The JSON data to be deserialized.
+
+    Returns:
+        (object): The deserialized JSON data.
+
+    Raises:
+        synapse.exc.BadJsonText: This exception is raised when there is an error
+            deserializing the provided data.
+    '''
+    try:
+        return orjson.loads(s)
+    except orjson.JSONDecodeError as exc:
+        extra = {'synapse': {'fn': 'loads', 'reason': str(exc)}}
+        logger.warning('Using fallback JSON deserialization. Please report this to Vertex.', extra=extra)
+        return _fallback_loads(s)
+
+def load(fp: BinaryIO) -> Any:
+    '''
+    Deserialize JSON data from a file.
+
+    Similar to the standard library json.load().
+
+    Arguments:
+        fp (file): The python file pointer to read the data from.
+
+    Returns:
+        (object): The deserialized JSON data.
+
+    Raises:
+        synapse.exc.BadJsonText: This exception is raised when there is an error
+            deserializing the provided data.
+    '''
+    return loads(fp.read())
+
+def _fallback_dumps(obj: Any, sort_keys: bool = False, indent: bool = False, default: Optional[Callable] = None) -> bytes:
+    indent = 2 if indent else None
+
+    try:
+        ret = json.dumps(obj, sort_keys=sort_keys, indent=indent, default=default)
+        return ret.encode()
+    except TypeError as exc:
+        raise s_exc.MustBeJsonSafe(mesg=exc.args[0])
+
+def dumps(obj: Any, sort_keys: bool = False, indent: bool = False, default: Optional[Callable] = None, newline: bool = False) -> bytes:
+    '''
+    Serialize a python object to byte string.
+
+    Similar to the standard library json.dumps().
+
+    Arguments:
+        obj (object): The python object to serialize.
+        sort_keys (bool): Sort dictionary keys. Default: False.
+        indent (bool): Include 2 spaces of indentation. Default: False.
+        default (Optional[Callable]): Callback for serializing unknown object types. Default: None.
+        newline (bool): Append a newline to the end of the serialized data. Default: False.
+
+    Returns:
+        (bytes): The JSON serialized python object.
+
+    Raises:
+        synapse.exc.MustBeJsonSafe: This exception is raised when a python object cannot be serialized.
+    '''
+    opts = 0
+
+    if indent:
+        opts |= orjson.OPT_INDENT_2
+
+    if sort_keys:
+        opts |= orjson.OPT_SORT_KEYS
+
+    if newline:
+        opts |= orjson.OPT_APPEND_NEWLINE
+
+    try:
+        return orjson.dumps(obj, option=opts, default=default)
+
+    except orjson.JSONEncodeError as exc:
+        if not isinstance(exc.__cause__, UnicodeEncodeError):
+            raise s_exc.MustBeJsonSafe(mesg=exc.args[0])
+
+        extra = {'synapse': {'fn': 'dumps', 'reason': str(exc)}}
+        logger.warning('Using fallback JSON serialization. Please report this to Vertex.', extra=extra)
+
+        ret = _fallback_dumps(obj, sort_keys=sort_keys, indent=indent, default=default)
```

Comment:
The `indent` method of the fallback is implicitly typed as ` int | None`. Passing a `true` value will indent the output to `1` space. We should pass indent as either `None` or `2`, in to ensure that indented output has the same indentation that orjson provides.

Reply:
See line 63 (the first line in `_fallback_dumps()`) where it ensures it's either 2 or None.